gax/src/main/java/com/google/api/gax/rpc/Watchdog.java

```diff
@@ -0,0 +1,314 @@
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.rpc;
+
+import com.google.api.core.ApiClock;
+import com.google.api.core.InternalApi;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import java.util.Iterator;
+import java.util.Map.Entry;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.concurrent.GuardedBy;
+import org.threeten.bp.Duration;
+
+/**
+ * Prevents the streams from hanging indefinitely. This middleware garbage collects idle streams in
+ * case the user forgot to close a ServerStream or if a connection is reset and GRPC does not get
+ * notified.
+ *
+ * <p>For every {@code checkInterval}, this class checks two thresholds:
+ *
+ * <ul>
+ *   <li>waitingTimeout: the amount of time to wait for a response (after the caller signaled
+ *       demand) before forcefully closing the stream. Duration.ZERO disables the timeout.
+ *   <li>idleTimeout: the amount of time to wait before assuming that the caller forgot to close the
+ *       stream and forcefully closing the stream. This is measured from the last time the caller
+ *       had no outstanding demand. Duration.ZERO disables the timeout.
+ * </ul>
+ *
+ * @param <ResponseT> The type of the response.
+ */
+@InternalApi
+public class Watchdog<ResponseT> {
+  // Dummy value to convert the ConcurrentHashMap into a Set
+  private static Object VALUE_MARKER = new Object();
+  private final ConcurrentHashMap<WatchdogStream, Object> openStreams = new ConcurrentHashMap<>();
+
+  private final ScheduledExecutorService executor;
+  private final ApiClock clock;
+  private final Duration checkInterval;
+  private final Duration idleTimeout;
+
+  public Watchdog(
+      ScheduledExecutorService executor,
+      ApiClock clock,
+      Duration checkInterval,
+      Duration idleTimeout) {
+
+    Preconditions.checkNotNull(executor, "executor can't be null");
```

Comment:
Minor: throwing NullPointerException with stacktrace should be enough to figure out what was null. We usually do not have explicit error message for this, so `Preconditions.checkNotNull(executor)` should be just enough here (same for other non-null params).

Reply:
I would prefer to keep it, it's not hurting anything and will save someone a trip to the source to figure out which param was null.